In [27]:
import pandas as pd
import numpy as np
import matplotlib as plt
import seaborn as sns

In [28]:
%matplotlib inline
pd.set_option('display.max_columns', None)

In [29]:
# import excel file Nov 2018 General results.xlsx
tn_2018g=pd.read_excel('../data/2018_election/Nov 2018 General results.xlsx',
                      usecols=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE','RNAME1','PARTY1','PVTALLY1',
                             'RNAME2','PARTY2','PVTALLY2','RNAME3','PARTY3','PVTALLY3','RNAME4','PARTY4','PVTALLY4',
                             'RNAME5','PARTY5','PVTALLY5','RNAME6','PARTY6','PVTALLY6','RNAME7','PARTY7','PVTALLY7',
                             'RNAME8','PARTY8','PVTALLY8','RNAME9','PARTY9','PVTALLY9','RNAME10','PARTY10','PVTALLY10'])
tn_2018g

COUNTY       PRECINCT  \
0      Anderson  Andersonville   
1      Anderson  Andersonville   
2      Anderson  Andersonville   
3      Anderson  Andersonville   
4      Anderson  Andersonville   
...         ...            ...   
13237    Wilson    Provisional   
13238    Wilson    Provisional   
13239    Wilson    Provisional   
13240    Wilson    Provisional   
13241    Wilson    Provisional   

                                              OFFICENAME   ELECTDATE  \
0                                               Governor  2018-11-06   
1                                               Governor  2018-11-06   
2                                               Governor  2018-11-06   
3                                   United States Senate  2018-11-06   
4      United States House of Representatives District 3  2018-11-06   
...                                                  ...         ...   
13237                               United States Senate  2018-11-06   
13238  United States House of Representatives District 6  2018-11-06   
13239                       Tennessee Senate District 17  2018-11-06   
13240     Tennessee House of Representatives District 46  2018-11-06   
13241     Tennessee House of Representatives District 57  2018-11-06   

                       RNAME1       PARTY1  PVTALLY1              RNAME2  \
0                    Bill Lee   Republican      1011           Karl Dean   
1          Tommy Ray McAnally  Independent         0  Jessie D. McDonald   
2      Tracy C. Yaste Tisdale  Independent         0          Mike Toews   
3            Marsha Blackburn   Republican       954       Phil Bredesen   
4           Chuck Fleischmann   Republican      1010   Danielle Mitchell   
...                       ...          ...       ...                 ...   
13237        Marsha Blackburn   Republican        54       Phil Bredesen   
13238               John Rose   Republican        54         Dawn Barlow   
13239               Mark Pody   Republican        61    Mary Alice Carfi   
13240              Clark Boyd   Republican        18          Mark Cagle   
13241            Susan M Lynn   Republican        37         Jordan Cole   

            PARTY2  PVTALLY2                  RNAME3       PARTY3  PVTALLY3  \
0       Democratic     244.0    Mark CoonRippy Brown  Independent       2.0   
1      Independent       0.0  Toney Randall Mitchell  Independent       0.0   
2      Independent       0.0              Rick Tyler  Independent       0.0   
3       Democratic     289.0         Trudy A. Austin  Independent       5.0   
4       Democratic     215.0              Rick Tyler  Independent      16.0   
...            ...       ...                     ...          ...       ...   
13237   Democratic      35.0         Trudy A. Austin  Independent       0.0   
13238   Democratic      30.0              Lloyd Dunn  Independent       3.0   
13239   Democratic      28.0                     NaN          NaN       NaN   
13240   Democratic      10.0                     NaN          NaN       NaN   
13241   Democratic      21.0                     NaN          NaN       NaN   

                RNAME4       PARTY4  PVTALLY4               RNAME5  \
0      Sherry L. Clark  Independent       4.0       Justin Cornett   
1       Yvonne Neubert  Independent       0.0  Alfred Shawn Rapoza   
2      Vinnie Vineyard  Independent       0.0     Jaron D. Weidner   
3          John Carico  Independent       5.0            Dean Hill   
4                  NaN          NaN       NaN                  NaN   
...                ...          ...       ...                  ...   
13237      John Carico  Independent       1.0            Dean Hill   
13238       David Ross  Independent       2.0                  NaN   
13239              NaN          NaN       NaN                  NaN   
13240              NaN          NaN       NaN                  NaN   
13241              NaN          NaN       NaN                  NaN   

            PARTY5  PVTALLY5    

In [30]:
tn_2018g.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13242 entries, 0 to 13241
Data columns (total 34 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   COUNTY      13242 non-null  object 
 1   PRECINCT    13242 non-null  object 
 2   OFFICENAME  13242 non-null  object 
 3   ELECTDATE   13242 non-null  object 
 4   RNAME1      13242 non-null  object 
 5   PARTY1      13242 non-null  object 
 6   PVTALLY1    13242 non-null  int64  
 7   RNAME2      12606 non-null  object 
 8   PARTY2      12483 non-null  object 
 9   PVTALLY2    12606 non-null  float64
 10  RNAME3      10363 non-null  object 
 11  PARTY3      10103 non-null  object 
 12  PVTALLY3    10363 non-null  float64
 13  RNAME4      8967 non-null   object 
 14  PARTY4      8677 non-null   object 
 15  PVTALLY4    8967 non-null   float64
 16  RNAME5      8130 non-null   object 
 17  PARTY5      8130 non-null   object 
 18  PVTALLY5    8130 non-null   float64
 19  RNAME6      8130 non-null

In [31]:
# merge entries in all RNAME columns into new column "name"
tn_2018g_namemelt=tn_2018g.melt(id_vars=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE'],value_vars=['RNAME1','RNAME2','RNAME3','RNAME4','RNAME5','RNAME6','RNAME7','RNAME8','RNAME9','RNAME10'],value_name="name")
# drop column "variable" created by .melt
tn_2018g_namemelt=tn_2018g_namemelt.drop(['variable'],axis=1)                          
tn_2018g_namemelt.head(9405)

COUNTY       PRECINCT  \
0     Anderson  Andersonville   
1     Anderson  Andersonville   
2     Anderson  Andersonville   
3     Anderson  Andersonville   
4     Anderson  Andersonville   
...        ...            ...   
9400     Roane      Lawnville   
9401     Roane         Renfro   
9402     Roane         Renfro   
9403     Roane         Renfro   
9404     Roane         Renfro   

                                             OFFICENAME   ELECTDATE  \
0                                              Governor  2018-11-06   
1                                              Governor  2018-11-06   
2                                              Governor  2018-11-06   
3                                  United States Senate  2018-11-06   
4     United States House of Representatives District 3  2018-11-06   
...                                                 ...         ...   
9400     Tennessee House of Representatives District 32  2018-11-06   
9401                                           Governor  2018-11-06   
9402                                           Governor  2018-11-06   
9403                                           Governor  2018-11-06   
9404                               United States Senate  2018-11-06   

                        name  
0                   Bill Lee  
1         Tommy Ray McAnally  
2     Tracy C. Yaste Tisdale  
3           Marsha Blackburn  
4          Chuck Fleischmann  
...                      ...  
9400             Kent Calfee  
9401                Bill Lee  
9402      Tommy Ray McAnally  
9403  Tracy C. Yaste Tisdale  
9404        Marsha Blackburn  

[9405 rows x 5 columns]

In [32]:
# merge entries in all PARTY columns into new column "party"
tn_2018g_partymelt=tn_2018g.melt(id_vars=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE'],value_vars=['PARTY1','PARTY2','PARTY3','PARTY4','PARTY5','PARTY6','PARTY7','PARTY8','PARTY9','PARTY10'],value_name="party")
# drop column "variable" created by .melt
tn_2018g_partymelt=tn_2018g_partymelt.drop(['variable'],axis=1)
tn_2018g_partymelt.head(9405)

COUNTY       PRECINCT  \
0     Anderson  Andersonville   
1     Anderson  Andersonville   
2     Anderson  Andersonville   
3     Anderson  Andersonville   
4     Anderson  Andersonville   
...        ...            ...   
9400     Roane      Lawnville   
9401     Roane         Renfro   
9402     Roane         Renfro   
9403     Roane         Renfro   
9404     Roane         Renfro   

                                             OFFICENAME   ELECTDATE  \
0                                              Governor  2018-11-06   
1                                              Governor  2018-11-06   
2                                              Governor  2018-11-06   
3                                  United States Senate  2018-11-06   
4     United States House of Representatives District 3  2018-11-06   
...                                                 ...         ...   
9400     Tennessee House of Representatives District 32  2018-11-06   
9401                                           Governor  2018-11-06   
9402                                           Governor  2018-11-06   
9403                                           Governor  2018-11-06   
9404                               United States Senate  2018-11-06   

            party  
0      Republican  
1     Independent  
2     Independent  
3      Republican  
4      Republican  
...           ...  
9400   Republican  
9401   Republican  
9402  Independent  
9403  Independent  
9404   Republican  

[9405 rows x 5 columns]

In [33]:
# merge entries in all PVTALLY columns into new column "vote_count"
tn_2018g_pvtallymelt=tn_2018g.melt(id_vars=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE'],value_vars=['PVTALLY1','PVTALLY2','PVTALLY3','PVTALLY4','PVTALLY5','PVTALLY6','PVTALLY7','PVTALLY8','PVTALLY9','PVTALLY10'],value_name="vote_count")
# drop column "variable" created by .melt
tn_2018g_pvtallymelt=tn_2018g_pvtallymelt.drop(['variable'],axis=1)
tn_2018g_pvtallymelt.head(9405)

COUNTY       PRECINCT  \
0     Anderson  Andersonville   
1     Anderson  Andersonville   
2     Anderson  Andersonville   
3     Anderson  Andersonville   
4     Anderson  Andersonville   
...        ...            ...   
9400     Roane      Lawnville   
9401     Roane         Renfro   
9402     Roane         Renfro   
9403     Roane         Renfro   
9404     Roane         Renfro   

                                             OFFICENAME   ELECTDATE  \
0                                              Governor  2018-11-06   
1                                              Governor  2018-11-06   
2                                              Governor  2018-11-06   
3                                  United States Senate  2018-11-06   
4     United States House of Representatives District 3  2018-11-06   
...                                                 ...         ...   
9400     Tennessee House of Representatives District 32  2018-11-06   
9401                                           Governor  2018-11-06   
9402                                           Governor  2018-11-06   
9403                                           Governor  2018-11-06   
9404                               United States Senate  2018-11-06   

      vote_count  
0         1011.0  
1            0.0  
2            0.0  
3          954.0  
4         1010.0  
...          ...  
9400       995.0  
9401       488.0  
9402         0.0  
9403         0.0  
9404       466.0  

[9405 rows x 5 columns]

In [34]:
# add "party" and "vote_count" columns to tn_2016_namemelt
party=tn_2018g_partymelt['party']
pvtally=tn_2018g_pvtallymelt['vote_count']
tn_2018g_namemelt=tn_2018g_namemelt.join(party).join(pvtally)
tn_2018g_namemelt.head(3)

COUNTY       PRECINCT OFFICENAME   ELECTDATE                    name  \
0  Anderson  Andersonville   Governor  2018-11-06                Bill Lee   
1  Anderson  Andersonville   Governor  2018-11-06      Tommy Ray McAnally   
2  Anderson  Andersonville   Governor  2018-11-06  Tracy C. Yaste Tisdale   

         party  vote_count  
0   Republican      1011.0  
1  Independent         0.0  
2  Independent         0.0

In [35]:
tn_2018g_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 132420 entries, 0 to 132419
Data columns (total 7 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   COUNTY      132420 non-null  object 
 1   PRECINCT    132420 non-null  object 
 2   OFFICENAME  132420 non-null  object 
 3   ELECTDATE   132420 non-null  object 
 4   name        87282 non-null   object 
 5   party       84621 non-null   object 
 6   vote_count  87282 non-null   float64
dtypes: float64(1), object(6)
memory usage: 7.1+ MB


In [36]:
# lower case column names
tn_2018g_namemelt.columns=tn_2018g_namemelt.columns.str.lower()
tn_2018g_namemelt

county       precinct  \
0       Anderson  Andersonville   
1       Anderson  Andersonville   
2       Anderson  Andersonville   
3       Anderson  Andersonville   
4       Anderson  Andersonville   
...          ...            ...   
132415    Wilson    Provisional   
132416    Wilson    Provisional   
132417    Wilson    Provisional   
132418    Wilson    Provisional   
132419    Wilson    Provisional   

                                               officename   electdate  \
0                                                Governor  2018-11-06   
1                                                Governor  2018-11-06   
2                                                Governor  2018-11-06   
3                                    United States Senate  2018-11-06   
4       United States House of Representatives District 3  2018-11-06   
...                                                   ...         ...   
132415                               United States Senate  2018-11-06   
132416  United States House of Representatives District 6  2018-11-06   
132417                       Tennessee Senate District 17  2018-11-06   
132418     Tennessee House of Representatives District 46  2018-11-06   
132419     Tennessee House of Representatives District 57  2018-11-06   

                          name        party  vote_count  
0                     Bill Lee   Republican      1011.0  
1           Tommy Ray McAnally  Independent         0.0  
2       Tracy C. Yaste Tisdale  Independent         0.0  
3             Marsha Blackburn   Republican       954.0  
4            Chuck Fleischmann   Republican      1010.0  
...                        ...          ...         ...  
132415                     NaN          NaN         NaN  
132416                     NaN          NaN         NaN  
132417                     NaN          NaN         NaN  
132418                     NaN          NaN         NaN  
132419                     NaN          NaN         NaN  

[132420 rows x 7 columns]

In [37]:
# change null values in "party" column to "no party"
tn_2018g_namemelt['party'].fillna("No party", inplace = True)
tn_2018g_namemelt

county       precinct  \
0       Anderson  Andersonville   
1       Anderson  Andersonville   
2       Anderson  Andersonville   
3       Anderson  Andersonville   
4       Anderson  Andersonville   
...          ...            ...   
132415    Wilson    Provisional   
132416    Wilson    Provisional   
132417    Wilson    Provisional   
132418    Wilson    Provisional   
132419    Wilson    Provisional   

                                               officename   electdate  \
0                                                Governor  2018-11-06   
1                                                Governor  2018-11-06   
2                                                Governor  2018-11-06   
3                                    United States Senate  2018-11-06   
4       United States House of Representatives District 3  2018-11-06   
...                                                   ...         ...   
132415                               United States Senate  2018-11-06   
132416  United States House of Representatives District 6  2018-11-06   
132417                       Tennessee Senate District 17  2018-11-06   
132418     Tennessee House of Representatives District 46  2018-11-06   
132419     Tennessee House of Representatives District 57  2018-11-06   

                          name        party  vote_count  
0                     Bill Lee   Republican      1011.0  
1           Tommy Ray McAnally  Independent         0.0  
2       Tracy C. Yaste Tisdale  Independent         0.0  
3             Marsha Blackburn   Republican       954.0  
4            Chuck Fleischmann   Republican      1010.0  
...                        ...          ...         ...  
132415                     NaN     No party         NaN  
132416                     NaN     No party         NaN  
132417                     NaN     No party         NaN  
132418                     NaN     No party         NaN  
132419                     NaN     No party         NaN  

[132420 rows x 7 columns]

In [39]:
tn_2018g_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 132420 entries, 0 to 132419
Data columns (total 7 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   county      132420 non-null  object 
 1   precinct    132420 non-null  object 
 2   officename  132420 non-null  object 
 3   electdate   132420 non-null  object 
 4   name        87282 non-null   object 
 5   party       132420 non-null  object 
 6   vote_count  87282 non-null   float64
dtypes: float64(1), object(6)
memory usage: 7.1+ MB


In [41]:
# change null values in "vote_count" column to "0"
tn_2018g_namemelt['vote_count'].fillna(0)

0         1011.0
1            0.0
2            0.0
3          954.0
4         1010.0
           ...  
132415       0.0
132416       0.0
132417       0.0
132418       0.0
132419       0.0
Name: vote_count, Length: 132420, dtype: float64

In [43]:
tn_2018g_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 132420 entries, 0 to 132419
Data columns (total 7 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   county      132420 non-null  object 
 1   precinct    132420 non-null  object 
 2   officename  132420 non-null  object 
 3   electdate   132420 non-null  object 
 4   name        87282 non-null   object 
 5   party       132420 non-null  object 
 6   vote_count  87282 non-null   float64
dtypes: float64(1), object(6)
memory usage: 7.1+ MB


### 87282 non-null names and 87282 non-null vote counts
### Do the nulls in "name" correlate one to one with "no party"?

In [44]:
tn_2018g_namemelt[tn_2018g_namemelt['name'].isna()]

county           precinct  \
13877   Bradley       1-3 Hopewell   
13891   Bradley  2-2 Walker Valley   
13913   Bradley      4-1 Oak Grove   
13920   Bradley   4-2 Michigan Ave   
13941   Bradley    6-1 Valley View   
...         ...                ...   
132415   Wilson        Provisional   
132416   Wilson        Provisional   
132417   Wilson        Provisional   
132418   Wilson        Provisional   
132419   Wilson        Provisional   

                                               officename   electdate name  \
13877      Tennessee House of Representatives District 22  2018-11-06  NaN   
13891      Tennessee House of Representatives District 22  2018-11-06  NaN   
13913      Tennessee House of Representatives District 22  2018-11-06  NaN   
13920      Tennessee House of Representatives District 22  2018-11-06  NaN   
13941      Tennessee House of Representatives District 22  2018-11-06  NaN   
...                                                   ...         ...  ...   
132415                               United States Senate  2018-11-06  NaN   
132416  United States House of Representatives District 6  2018-11-06  NaN   
132417                       Tennessee Senate District 17  2018-11-06  NaN   
132418     Tennessee House of Representatives District 46  2018-11-06  NaN   
132419     Tennessee House of Representatives District 57  2018-11-06  NaN   

           party  vote_count  
13877   No party         NaN  
13891   No party         NaN  
13913   No party         NaN  
13920   No party         NaN  
13941   No party         NaN  
...          ...         ...  
132415  No party         NaN  
132416  No party         NaN  
132417  No party         NaN  
132418  No party         NaN  
132419  No party         NaN  

[45138 rows x 7 columns]

In [45]:
tn_2018g_namemelt[tn_2018g_namemelt['name'].isna()].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 45138 entries, 13877 to 132419
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   county      45138 non-null  object 
 1   precinct    45138 non-null  object 
 2   officename  45138 non-null  object 
 3   electdate   45138 non-null  object 
 4   name        0 non-null      object 
 5   party       45138 non-null  object 
 6   vote_count  0 non-null      float64
dtypes: float64(1), object(6)
memory usage: 2.8+ MB


### Looks safe to drop rows with no votes tallied

In [46]:
# drop rows with no name and vote count
tn_2018g_final=tn_2018g_namemelt.dropna()

In [47]:
tn_2018g_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 87282 entries, 0 to 132413
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   county      87282 non-null  object 
 1   precinct    87282 non-null  object 
 2   officename  87282 non-null  object 
 3   electdate   87282 non-null  object 
 4   name        87282 non-null  object 
 5   party       87282 non-null  object 
 6   vote_count  87282 non-null  float64
dtypes: float64(1), object(6)
memory usage: 5.3+ MB


In [48]:
# convert vote_count data from float64 to int 32 to to drop ".0"
tn_2018g_final=tn_2018g_final.astype({'vote_count':'int32'})
tn_2018g_final

county       precinct  \
0       Anderson  Andersonville   
1       Anderson  Andersonville   
2       Anderson  Andersonville   
3       Anderson  Andersonville   
4       Anderson  Andersonville   
...          ...            ...   
132397    Wilson           25-2   
132404    Wilson       Absentee   
132405    Wilson       Absentee   
132412    Wilson    Provisional   
132413    Wilson    Provisional   

                                               officename   electdate  \
0                                                Governor  2018-11-06   
1                                                Governor  2018-11-06   
2                                                Governor  2018-11-06   
3                                    United States Senate  2018-11-06   
4       United States House of Representatives District 3  2018-11-06   
...                                                   ...         ...   
132397                                           Governor  2018-11-06   
132404                                           Governor  2018-11-06   
132405                                           Governor  2018-11-06   
132412                                           Governor  2018-11-06   
132413                                           Governor  2018-11-06   

                           name        party  vote_count  
0                      Bill Lee   Republican        1011  
1            Tommy Ray McAnally  Independent           0  
2        Tracy C. Yaste Tisdale  Independent           0  
3              Marsha Blackburn   Republican         954  
4             Chuck Fleischmann   Republican        1010  
...                         ...          ...         ...  
132397  Jeremy Allen Stephenson  Independent           0  
132404             Matthew Koch  Independent           0  
132405  Jeremy Allen Stephenson  Independent           0  
132412             Matthew Koch  Independent           0  
132413  Jeremy Allen Stephenson  Independent           0  

[87282 rows x 7 columns]

In [49]:
tn_2018g_final.dtypes

county        object
precinct      object
officename    object
electdate     object
name          object
party         object
vote_count     int32
dtype: object

In [50]:
tn_2018g_final['officename'].value_counts()

Governor                                             57652
United States Senate                                 15904
United States House of Representatives District 7     1124
United States House of Representatives District 2     1068
United States House of Representatives District 6     1064
                                                     ...  
Tennessee House of Representatives District 60          18
Tennessee House of Representatives District 88          16
Tennessee House of Representatives District 93          16
Tennessee House of Representatives District 1           11
Tennessee House of Representatives District 81           9
Name: officename, Length: 128, dtype: int64

In [51]:
tn_2018g_final['officename'].unique()

array(['Governor', 'United States Senate',
       'United States House of Representatives District 3',
       'Tennessee Senate District 5',
       'Tennessee House of Representatives District 36',
       'Tennessee House of Representatives District 33',
       'United States House of Representatives District 4',
       'Tennessee House of Representatives District 62',
       'United States House of Representatives District 7',
       'Tennessee House of Representatives District 75',
       'United States House of Representatives District 8',
       'Tennessee Senate District 15',
       'Tennessee House of Representatives District 31',
       'United States House of Representatives District 2',
       'Tennessee Senate District 2 (unexpired term)',
       'Tennessee House of Representatives District 8',
       'Tennessee House of Representatives District 20',
       'Tennessee House of Representatives District 24',
       'Tennessee Senate District 9',
       'Tennessee House of Repre

In [52]:
tn_2018g_final['vote_count'].value_counts()

0       41700
1       12006
2        5343
3        2925
4        1791
        ...  
1708        1
2254        1
1900        1
2126        1
2047        1
Name: vote_count, Length: 2155, dtype: int64

### 41700 rows have a vote count of 0. Drop those rows

In [53]:
tn_2018g_final=tn_2018g_final[tn_2018g_final['vote_count'] != 0]
tn_2018g_final=tn_2018g_final.reset_index(drop=True)
tn_2018g_final.head()

county       precinct                                         officename  \
0  Anderson  Andersonville                                           Governor   
1  Anderson  Andersonville                               United States Senate   
2  Anderson  Andersonville  United States House of Representatives District 3   
3  Anderson  Andersonville                        Tennessee Senate District 5   
4  Anderson  Andersonville     Tennessee House of Representatives District 36   

    electdate               name       party  vote_count  
0  2018-11-06           Bill Lee  Republican        1011  
1  2018-11-06   Marsha Blackburn  Republican         954  
2  2018-11-06  Chuck Fleischmann  Republican        1010  
3  2018-11-06      Randy McNally  Republican        1051  
4  2018-11-06   Dennis H. Powers  Republican         996

In [54]:
# confirm no 0's in vote count
tn_2018g_final['vote_count'].value_counts()

1        12006
2         5343
3         2925
4         1791
5         1207
         ...  
11752        1
1389         1
9545         1
3340         1
2049         1
Name: vote_count, Length: 2154, dtype: int64

In [56]:
# create data frame tn_2018_gov for presidental election
tn_2018_gov_int=tn_2018g_final[tn_2018g_final['officename'].str.match('Governor')]
tn_2018_gov_int=tn_2018_gov_int.reset_index(drop=True)
tn_2018_gov_int

county       precinct officename   electdate  \
0      Anderson  Andersonville   Governor  2018-11-06   
1      Anderson     Briceville   Governor  2018-11-06   
2      Anderson     Briceville   Governor  2018-11-06   
3      Anderson       Bull Run   Governor  2018-11-06   
4      Anderson       Bull Run   Governor  2018-11-06   
...         ...            ...        ...         ...   
20354    Wilson           17-1   Governor  2018-11-06   
20355    Wilson           18-1   Governor  2018-11-06   
20356    Wilson           21-1   Governor  2018-11-06   
20357    Wilson           23-1   Governor  2018-11-06   
20358    Wilson           24-1   Governor  2018-11-06   

                          name        party  vote_count  
0                     Bill Lee   Republican        1011  
1                     Bill Lee   Republican         181  
2       Tracy C. Yaste Tisdale  Independent           1  
3                     Bill Lee   Republican        1177  
4           Tommy Ray McAnally  Independent           1  
...                        ...          ...         ...  
20354  Jeremy Allen Stephenson  Independent           1  
20355             Matthew Koch  Independent           1  
20356  Jeremy Allen Stephenson  Independent           1  
20357             Matthew Koch  Independent           1  
20358  Jeremy Allen Stephenson  Independent           1  

[20359 rows x 7 columns]

In [57]:
tn_2018_gov_int.head(20)

county        precinct officename   electdate                    name  \
0   Anderson   Andersonville   Governor  2018-11-06                Bill Lee   
1   Anderson      Briceville   Governor  2018-11-06                Bill Lee   
2   Anderson      Briceville   Governor  2018-11-06  Tracy C. Yaste Tisdale   
3   Anderson        Bull Run   Governor  2018-11-06                Bill Lee   
4   Anderson        Bull Run   Governor  2018-11-06      Tommy Ray McAnally   
5   Anderson         Clinton   Governor  2018-11-06                Bill Lee   
6   Anderson         Clinton   Governor  2018-11-06  Tracy C. Yaste Tisdale   
7   Anderson    Clinton High   Governor  2018-11-06                Bill Lee   
8   Anderson         Claxton   Governor  2018-11-06                Bill Lee   
9   Anderson         Claxton   Governor  2018-11-06      Tommy Ray McAnally   
10  Anderson  Clinton Middle   Governor  2018-11-06                Bill Lee   
11  Anderson    Dutch Valley   Governor  2018-11-06                Bill Lee   
12  Anderson    Dutch Valley   Governor  2018-11-06      Tommy Ray McAnally   
13  Anderson    Dutch Valley   Governor  2018-11-06  Tracy C. Yaste Tisdale   
14  Anderson    Emory Valley   Governor  2018-11-06                Bill Lee   
15  Anderson    Emory Valley   Governor  2018-11-06      Tommy Ray McAnally   
16  Anderson        Fairview   Governor  2018-11-06                Bill Lee   
17  Anderson     Glen Alpine   Governor  2018-11-06                Bill Lee   
18  Anderson     Glen Alpine   Governor  2018-11-06      Tommy Ray McAnally   
19  Anderson     Glen Alpine   Governor  2018-11-06  Tracy C. Yaste Tisdale   

          party  vote_count  
0    Republican        1011  
1    Republican         181  
2   Independent           1  
3    Republican        1177  
4   Independent           1  
5    Republican        1170  
6   Independent           1  
7    Republican         676  
8    Republican        1082  
9   Independent           1  
10   Republican         356  
11   Republican         219  
12  Independent           1  
13  Independent           1  
14   Republican         971  
15  Independent           1  
16   Republican         472  
17   Republican         915  
18  Independent           2  
19  Independent           1

### Add column that has total vote count by count for use in calculating county vote percentages

In [64]:
# group county precincts into county per candidate, in descending order of votes within county
tn_2018_gov_int=tn_2018_gov_int.groupby(['electdate','officename','county','name','party'], as_index=False)['vote_count'].sum()
tn_2018_gov=tn_2018_gov_int.sort_values(by=['county','vote_count'],ascending=False).reset_index(drop=True)
tn_2018_gov.head(30)

electdate officename      county                        name  \
0   2018-11-06   Governor      Wilson                    Bill Lee   
1   2018-11-06   Governor      Wilson                   Karl Dean   
2   2018-11-06   Governor      Wilson             Sherry L. Clark   
3   2018-11-06   Governor      Wilson                 Mark Wright   
4   2018-11-06   Governor      Wilson               Heather Scott   
5   2018-11-06   Governor      Wilson        Mark CoonRippy Brown   
6   2018-11-06   Governor      Wilson                  Mike Toews   
7   2018-11-06   Governor      Wilson                  Rick Tyler   
8   2018-11-06   Governor      Wilson            Patrick Whitlock   
9   2018-11-06   Governor      Wilson          Sean Bruce Fleming   
10  2018-11-06   Governor      Wilson             Vinnie Vineyard   
11  2018-11-06   Governor      Wilson              Joe B. Wilmoth   
12  2018-11-06   Governor      Wilson      Tracy C. Yaste Tisdale   
13  2018-11-06   Governor      Wilson                   Cory King   
14  2018-11-06   Governor      Wilson             Gabriel Fancher   
15  2018-11-06   Governor      Wilson                  Chad Riden   
16  2018-11-06   Governor      Wilson              Justin Cornett   
17  2018-11-06   Governor      Wilson            Jaron D. Weidner   
18  2018-11-06   Governor      Wilson  William Andrew Helmstetter   
19  2018-11-06   Governor      Wilson         Alfred Shawn Rapoza   
20  2018-11-06   Governor      Wilson   George Blackwell Smith IV   
21  2018-11-06   Governor      Wilson                Matthew Koch   
22  2018-11-06   Governor      Wilson              Yvonne Neubert   
23  2018-11-06   Governor      Wilson          Jessie D. McDonald   
24  2018-11-06   Governor      Wilson     Jeremy Allen Stephenson   
25  2018-11-06   Governor      Wilson      Toney Randall Mitchell   
26  2018-11-06   Governor      Wilson          Robert Sawyers Sr.   
27  2018-11-06   Governor      Wilson          Tommy Ray McAnally   
28  2018-11-06   Governor  Williamson                    Bill Lee   
29  2018-11-06   Governor  Williamson                   Karl Dean   

          party  vote_count  
0    Republican       35488  
1    Democratic       16987  
2   Independent          95  
3   Independent          79  
4   Independent          51  
5   Independent          46  
6   Independent          42  
7   Independent          33  
8   Independent          28  
9   Independent          25  
10  Independent          25  
11  Independent          18  
12  Independent          16  
13  Independent          14  
14  Independent          13  
15  Independent          12  
16  Independent          12  
17  Independent          10  
18  Independent          10  
19  Independent           9  
20  Independent           9  
21  Independent           9  
22  Independent           9  
23  Independent           8  
24  Independent           7  
25  Independent           5  
26  Independent           4  
27  Independent           3  
28   Republican       68189  
29   Democratic       34647

In [65]:
# create new dataframe "county_vote" to use to find sum of votes per county
county_vote=tn_2018_gov_int.loc[0:,['county','vote_count']]
county_vote

county  vote_count
0     Anderson           2
1     Anderson       16265
2     Anderson          13
3     Anderson           9
4     Anderson           4
...        ...         ...
2225    Wilson           5
2226    Wilson          16
2227    Wilson          25
2228    Wilson          10
2229    Wilson           9

[2230 rows x 2 columns]

In [66]:
#sum county votes
county_vote_sum=county_vote.groupby(['county'])['vote_count'].sum()
county_vote_sum

county
Anderson       26040
Bedford        13049
Benton          5346
Bledsoe         4072
Blount         46775
               ...  
Wayne           4731
Weakley         9892
White           8444
Williamson    105654
Wilson         53067
Name: vote_count, Length: 95, dtype: int32

In [67]:
# use "county_vote_sum" to add column "sum_county_vote" to tn_2018_gov correlating with value in each row
tn_2018_gov['sum_county_vote']=tn_2018_gov['county'].map(county_vote_sum)
tn_2018_gov.head(20)

electdate officename  county                        name        party  \
0   2018-11-06   Governor  Wilson                    Bill Lee   Republican   
1   2018-11-06   Governor  Wilson                   Karl Dean   Democratic   
2   2018-11-06   Governor  Wilson             Sherry L. Clark  Independent   
3   2018-11-06   Governor  Wilson                 Mark Wright  Independent   
4   2018-11-06   Governor  Wilson               Heather Scott  Independent   
5   2018-11-06   Governor  Wilson        Mark CoonRippy Brown  Independent   
6   2018-11-06   Governor  Wilson                  Mike Toews  Independent   
7   2018-11-06   Governor  Wilson                  Rick Tyler  Independent   
8   2018-11-06   Governor  Wilson            Patrick Whitlock  Independent   
9   2018-11-06   Governor  Wilson          Sean Bruce Fleming  Independent   
10  2018-11-06   Governor  Wilson             Vinnie Vineyard  Independent   
11  2018-11-06   Governor  Wilson              Joe B. Wilmoth  Independent   
12  2018-11-06   Governor  Wilson      Tracy C. Yaste Tisdale  Independent   
13  2018-11-06   Governor  Wilson                   Cory King  Independent   
14  2018-11-06   Governor  Wilson             Gabriel Fancher  Independent   
15  2018-11-06   Governor  Wilson                  Chad Riden  Independent   
16  2018-11-06   Governor  Wilson              Justin Cornett  Independent   
17  2018-11-06   Governor  Wilson            Jaron D. Weidner  Independent   
18  2018-11-06   Governor  Wilson  William Andrew Helmstetter  Independent   
19  2018-11-06   Governor  Wilson         Alfred Shawn Rapoza  Independent   

    vote_count  sum_county_vote  
0        35488            53067  
1        16987            53067  
2           95            53067  
3           79            53067  
4           51            53067  
5           46            53067  
6           42            53067  
7           33            53067  
8           28            53067  
9           25            53067  
10          25            53067  
11          18            53067  
12          16            53067  
13          14            53067  
14          13            53067  
15          12            53067  
16          12            53067  
17          10            53067  
18          10            53067  
19           9            53067

In [68]:
# create "pct_county_vote" column for tn_2016_pres
tn_2018_gov['pct_county_vote']=round(tn_2018_gov['vote_count']/tn_2018_gov['sum_county_vote']*100,2)
tn_2018_gov

electdate officename    county                 name        party  \
0     2018-11-06   Governor    Wilson             Bill Lee   Republican   
1     2018-11-06   Governor    Wilson            Karl Dean   Democratic   
2     2018-11-06   Governor    Wilson      Sherry L. Clark  Independent   
3     2018-11-06   Governor    Wilson          Mark Wright  Independent   
4     2018-11-06   Governor    Wilson        Heather Scott  Independent   
...          ...        ...       ...                  ...          ...   
2225  2018-11-06   Governor  Anderson           Mike Toews  Independent   
2226  2018-11-06   Governor  Anderson   Robert Sawyers Sr.  Independent   
2227  2018-11-06   Governor  Anderson  Alfred Shawn Rapoza  Independent   
2228  2018-11-06   Governor  Anderson       Joe B. Wilmoth  Independent   
2229  2018-11-06   Governor  Anderson     Patrick Whitlock  Independent   

      vote_count  sum_county_vote  pct_county_vote  
0          35488            53067            66.87  
1          16987            53067            32.01  
2             95            53067             0.18  
3             79            53067             0.15  
4             51            53067             0.10  
...          ...              ...              ...  
2225           3            26040             0.01  
2226           3            26040             0.01  
2227           2            26040             0.01  
2228           1            26040             0.00  
2229           1            26040             0.00  

[2230 rows x 8 columns]

In [69]:
# how many parties?
tn_2018_gov['party'].value_counts()

Independent    2036
Democratic       95
Republican       95
No party          4
Name: party, dtype: int64

### Create dataframe "tn_2018_gov_pct_county"  with percent values per party per county

In [74]:
# create new dataframe with reduced columns values only
tn_2018_gov_pct_county=tn_2018_gov.loc[0:,['electdate','officename','county','party','pct_county_vote']]
tn_2018_gov_pct_county

electdate officename    county        party  pct_county_vote
0     2018-11-06   Governor    Wilson   Republican            66.87
1     2018-11-06   Governor    Wilson   Democratic            32.01
2     2018-11-06   Governor    Wilson  Independent             0.18
3     2018-11-06   Governor    Wilson  Independent             0.15
4     2018-11-06   Governor    Wilson  Independent             0.10
...          ...        ...       ...          ...              ...
2225  2018-11-06   Governor  Anderson  Independent             0.01
2226  2018-11-06   Governor  Anderson  Independent             0.01
2227  2018-11-06   Governor  Anderson  Independent             0.01
2228  2018-11-06   Governor  Anderson  Independent             0.00
2229  2018-11-06   Governor  Anderson  Independent             0.00

[2230 rows x 5 columns]

In [82]:
# create new dataframe with Republican values only
tn_2018_g_rep_pct=tn_2018_gov_pct_county[tn_2018_gov_pct_county['party']== "Republican"].reset_index(drop=True)
tn_2018_g_rep_pct

electdate officename      county       party  pct_county_vote
0   2018-11-06   Governor      Wilson  Republican            66.87
1   2018-11-06   Governor  Williamson  Republican            64.54
2   2018-11-06   Governor       White  Republican            75.49
3   2018-11-06   Governor     Weakley  Republican            72.82
4   2018-11-06   Governor       Wayne  Republican            81.93
..         ...        ...         ...         ...              ...
90  2018-11-06   Governor      Blount  Republican            70.46
91  2018-11-06   Governor     Bledsoe  Republican            76.18
92  2018-11-06   Governor      Benton  Republican            69.14
93  2018-11-06   Governor     Bedford  Republican            72.72
94  2018-11-06   Governor    Anderson  Republican            62.46

[95 rows x 5 columns]

In [83]:
# rename 'pct_county_vote' to 'rep_pct' and drop "party" column to prepare for merge 
tn_2018_g_rep_pct=tn_2018_g_rep_pct.rename(columns={'pct_county_vote':'Republican'},inplace=False).drop(['party'], axis=1)
tn_2018_g_rep_pct

electdate officename      county  Republican
0   2018-11-06   Governor      Wilson       66.87
1   2018-11-06   Governor  Williamson       64.54
2   2018-11-06   Governor       White       75.49
3   2018-11-06   Governor     Weakley       72.82
4   2018-11-06   Governor       Wayne       81.93
..         ...        ...         ...         ...
90  2018-11-06   Governor      Blount       70.46
91  2018-11-06   Governor     Bledsoe       76.18
92  2018-11-06   Governor      Benton       69.14
93  2018-11-06   Governor     Bedford       72.72
94  2018-11-06   Governor    Anderson       62.46

[95 rows x 4 columns]

In [84]:
# create new dataframe with Democratic values only
tn_2018_g_dem_pct=tn_2018_gov_pct_county[tn_2018_gov_pct_county['party']== "Democratic"].reset_index(drop=True)
tn_2018_g_dem_pct

electdate officename      county       party  pct_county_vote
0   2018-11-06   Governor      Wilson  Democratic            32.01
1   2018-11-06   Governor  Williamson  Democratic            32.79
2   2018-11-06   Governor       White  Democratic            23.74
3   2018-11-06   Governor     Weakley  Democratic            25.85
4   2018-11-06   Governor       Wayne  Democratic            17.29
..         ...        ...         ...         ...              ...
90  2018-11-06   Governor      Blount  Democratic            28.57
91  2018-11-06   Governor     Bledsoe  Democratic            22.67
92  2018-11-06   Governor      Benton  Democratic            29.91
93  2018-11-06   Governor     Bedford  Democratic            26.42
94  2018-11-06   Governor    Anderson  Democratic            36.34

[95 rows x 5 columns]

In [85]:
# rename 'pct_county_vote' to 'dem_pct' and drop "party" to prepare for merge 
tn_2018_g_dem_pct=tn_2018_g_dem_pct.rename(columns={'pct_county_vote':'Democratic'},inplace=False).drop(['party'], axis=1)
tn_2018_g_dem_pct

electdate officename      county  Democratic
0   2018-11-06   Governor      Wilson       32.01
1   2018-11-06   Governor  Williamson       32.79
2   2018-11-06   Governor       White       23.74
3   2018-11-06   Governor     Weakley       25.85
4   2018-11-06   Governor       Wayne       17.29
..         ...        ...         ...         ...
90  2018-11-06   Governor      Blount       28.57
91  2018-11-06   Governor     Bledsoe       22.67
92  2018-11-06   Governor      Benton       29.91
93  2018-11-06   Governor     Bedford       26.42
94  2018-11-06   Governor    Anderson       36.34

[95 rows x 4 columns]

In [87]:
# create new dataframe with non Republican or Democratic values only
tn_2018_g_other_pct=tn_2018_gov_pct_county[(tn_2018_gov_pct_county['party']=="Independent")|(tn_2018_gov_pct_county['party']=="No party")].reset_index(drop=True)
tn_2018_g_other_pct

electdate officename    county        party  pct_county_vote
0     2018-11-06   Governor    Wilson  Independent             0.18
1     2018-11-06   Governor    Wilson  Independent             0.15
2     2018-11-06   Governor    Wilson  Independent             0.10
3     2018-11-06   Governor    Wilson  Independent             0.09
4     2018-11-06   Governor    Wilson  Independent             0.08
...          ...        ...       ...          ...              ...
2035  2018-11-06   Governor  Anderson  Independent             0.01
2036  2018-11-06   Governor  Anderson  Independent             0.01
2037  2018-11-06   Governor  Anderson  Independent             0.01
2038  2018-11-06   Governor  Anderson  Independent             0.00
2039  2018-11-06   Governor  Anderson  Independent             0.00

[2040 rows x 5 columns]

In [88]:
# sum by county
tn_2018_g_other_pct=tn_2018_g_other_pct.groupby(['county'],as_index=False)['pct_county_vote'].sum()
tn_2018_g_other_pct

county  pct_county_vote
0     Anderson             1.20
1      Bedford             0.87
2       Benton             0.98
3      Bledsoe             1.13
4       Blount             0.99
..         ...              ...
90       Wayne             0.76
91     Weakley             1.32
92       White             0.76
93  Williamson             2.68
94      Wilson             1.14

[95 rows x 2 columns]

In [89]:
# rename 'pct_county_vote' to 'other_pct' and drop "party" to prepare for merge 
tn_2018_g_other_pct=tn_2018_g_other_pct.rename(columns={'pct_county_vote':'Other'})
tn_2018_g_other_pct

county  Other
0     Anderson   1.20
1      Bedford   0.87
2       Benton   0.98
3      Bledsoe   1.13
4       Blount   0.99
..         ...    ...
90       Wayne   0.76
91     Weakley   1.32
92       White   0.76
93  Williamson   2.68
94      Wilson   1.14

[95 rows x 2 columns]

In [92]:
# join tn_2018_g_rep, tn_2018_g_dem_pct and tn_2018__other_pct to form tn_2018_g_county_pct
dem=tn_2018_g_dem_pct['Democratic']
other=tn_2018_g_other_pct['Other']
tn_2018_g_county_pct=tn_2018_g_rep_pct.join(dem).join(other)
tn_2018_g_county_pct

electdate officename      county  Republican  Democratic  Other
0   2018-11-06   Governor      Wilson       66.87       32.01   1.20
1   2018-11-06   Governor  Williamson       64.54       32.79   0.87
2   2018-11-06   Governor       White       75.49       23.74   0.98
3   2018-11-06   Governor     Weakley       72.82       25.85   1.13
4   2018-11-06   Governor       Wayne       81.93       17.29   0.99
..         ...        ...         ...         ...         ...    ...
90  2018-11-06   Governor      Blount       70.46       28.57   0.76
91  2018-11-06   Governor     Bledsoe       76.18       22.67   1.32
92  2018-11-06   Governor      Benton       69.14       29.91   0.76
93  2018-11-06   Governor     Bedford       72.72       26.42   2.68
94  2018-11-06   Governor    Anderson       62.46       36.34   1.14

[95 rows x 6 columns]

### Add column for Republican margin over Democrat for heat mapping in Tableau

In [96]:
tn_2018_g_county_pct['Rep_Margin']= tn_2018_g_county_pct['Republican']-tn_2018_g_county_pct['Democratic']
tn_2018_g_county_pct

electdate officename      county  Republican  Democratic  Other  \
0   2018-11-06   Governor      Wilson       66.87       32.01   1.20   
1   2018-11-06   Governor  Williamson       64.54       32.79   0.87   
2   2018-11-06   Governor       White       75.49       23.74   0.98   
3   2018-11-06   Governor     Weakley       72.82       25.85   1.13   
4   2018-11-06   Governor       Wayne       81.93       17.29   0.99   
..         ...        ...         ...         ...         ...    ...   
90  2018-11-06   Governor      Blount       70.46       28.57   0.76   
91  2018-11-06   Governor     Bledsoe       76.18       22.67   1.32   
92  2018-11-06   Governor      Benton       69.14       29.91   0.76   
93  2018-11-06   Governor     Bedford       72.72       26.42   2.68   
94  2018-11-06   Governor    Anderson       62.46       36.34   1.14   

    Rep_Margin  
0        34.86  
1        31.75  
2        51.75  
3        46.97  
4        64.64  
..         ...  
90       41.89  
91       53.51  
92       39.23  
93       46.30  
94       26.12  

[95 rows x 7 columns]

In [94]:
#read TN_county_index_shape.csv to add index from TN Counties shape file for joining in Tableau
county_index=pd.read_csv('../data/TN_county_index_shape.csv')
county_index

county  INDEX_NO
0     Anderson        79
1      Bedford        26
2       Benton        41
3      Bledsoe        31
4       Blount        52
..         ...       ...
90       Wayne         6
91     Weakley        56
92       White        67
93  Williamson        44
94      Wilson        65

[95 rows x 2 columns]

In [101]:
# join with tn_2016_county_pres_pct
tn_2018_g_county_pct=tn_2018_g_county_pct.merge(county_index,on="county")
tn_2018_g_county_pct.head(50)

electdate officename      county  Republican  Democratic  Other  \
0  2018-11-06   Governor      Wilson       66.87       32.01   1.20   
1  2018-11-06   Governor  Williamson       64.54       32.79   0.87   
2  2018-11-06   Governor       White       75.49       23.74   0.98   
3  2018-11-06   Governor     Weakley       72.82       25.85   1.13   
4  2018-11-06   Governor       Wayne       81.93       17.29   0.99   
5  2018-11-06   Governor  Washington       68.98       30.16   1.03   
6  2018-11-06   Governor      Warren       67.44       31.76   1.45   
7  2018-11-06   Governor   Van Buren       70.27       28.20   0.55   
8  2018-11-06   Governor       Union       77.11       21.69   1.00   
9  2018-11-06   Governor      Unicoi       79.57       19.51   0.95   
10 2018-11-06   Governor   Trousdale       65.60       33.23   0.53   
11 2018-11-06   Governor      Tipton       72.85       25.41   1.78   
12 2018-11-06   Governor      Sumner       67.32       31.89   1.20   
13 2018-11-06   Governor    Sullivan       76.42       22.76   0.79   
14 2018-11-06   Governor     Stewart       69.24       29.67   1.42   
15 2018-11-06   Governor       Smith       72.38       26.78   1.87   
16 2018-11-06   Governor      Shelby       36.74       60.48   0.78   
17 2018-11-06   Governor      Sevier       76.73       21.43   0.88   
18 2018-11-06   Governor  Sequatchie       75.24       23.68   5.62   
19 2018-11-06   Governor       Scott       81.53       17.52   0.75   
20 2018-11-06   Governor  Rutherford       57.55       41.53   2.15   
21 2018-11-06   Governor   Robertson       69.61       29.71   0.54   
22 2018-11-06   Governor       Roane       71.91       27.00   1.47   
23 2018-11-06   Governor        Rhea       78.58       20.50   1.46   
24 2018-11-06   Governor      Putnam       67.33       31.64   1.43   
25 2018-11-06   Governor        Polk       74.73       24.38   0.61   
26 2018-11-06   Governor     Pickett       73.95       25.66   1.33   
27 2018-11-06   Governor       Perry       72.37       26.73   1.23   
28 2018-11-06   Governor     Overton       68.54       30.63   1.09   
29 2018-11-06   Governor       Obion       77.52       21.43   0.93   
30 2018-11-06   Governor      Morgan       76.89       22.07   1.32   
31 2018-11-06   Governor       Moore       77.57       20.91   0.78   
32 2018-11-06   Governor  Montgomery       53.98       43.82   0.91   
33 2018-11-06   Governor      Monroe       76.90       22.12   1.24   
34 2018-11-06   Governor       Meigs       76.72       22.23   1.53   
35 2018-11-06   Governor     McNairy       75.86       22.33   2.27   
36 2018-11-06   Governor      McMinn       77.53       21.37   0.97   
37 2018-11-06   Governor       Maury       65.07       34.06   1.20   
38 2018-11-06   Governor    Marshall       70.52       28.82   0.86   
39 2018-11-06   Governor      Marion       68.54       30.40   1.06   
40 2018-11-06   Governor     Madison       56.99       41.88   0.99   
41 2018-11-06   Governor       Macon       79.68       19.55   0.99   
42 2018-11-06   Governor      Loudon       75.05       23.93   0.77   
43 2018-11-06   Governor     Lincoln       77.33       20.07   1.01   
44 2018-11-06   Governor       Lewis       73.40       25.91   0.95   
45 2018-11-06   Governor    Lawrence       75.91       23.53   1.20   
46 2018-11-06   Governor  Lauderdale       60.38       38.14   1.19   
47 2018-11-06   Governor        Lake       65.05       23.05  11.92   
48 2018-11-06   Governor        Knox       57.68       41.12   1.44   
49 2018-11-06   Governor     Johnson       81.62       17.19   0.60   

    Rep_Margin  INDEX_NO_x  INDEX_NO_y  INDEX_NO  
0        34.86          65          65        65  
1        31.75          44          44        44  
2        51.75          67          67        67  
3        46.97          56          56        56  
4        64.64           6           6         6  
5        38.82          92          92        92  
6        35.68          4

In [98]:
tn_2018_g_county_pct.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 95 entries, 0 to 94
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   electdate   95 non-null     object 
 1   officename  95 non-null     object 
 2   county      95 non-null     object 
 3   Republican  95 non-null     float64
 4   Democratic  95 non-null     float64
 5   Other       95 non-null     float64
 6   Rep_Margin  95 non-null     float64
 7   INDEX_NO    95 non-null     int64  
dtypes: float64(4), int64(1), object(3)
memory usage: 6.7+ KB


In [99]:
# covert "electdate" to datetime from object
tn_2018_g_county_pct['electdate']=pd.to_datetime(tn_2018_g_county_pct['electdate'],format='%Y-%m-%d')
tn_2018_g_county_pct

electdate officename      county  Republican  Democratic  Other  \
0  2018-11-06   Governor      Wilson       66.87       32.01   1.20   
1  2018-11-06   Governor  Williamson       64.54       32.79   0.87   
2  2018-11-06   Governor       White       75.49       23.74   0.98   
3  2018-11-06   Governor     Weakley       72.82       25.85   1.13   
4  2018-11-06   Governor       Wayne       81.93       17.29   0.99   
..        ...        ...         ...         ...         ...    ...   
90 2018-11-06   Governor      Blount       70.46       28.57   0.76   
91 2018-11-06   Governor     Bledsoe       76.18       22.67   1.32   
92 2018-11-06   Governor      Benton       69.14       29.91   0.76   
93 2018-11-06   Governor     Bedford       72.72       26.42   2.68   
94 2018-11-06   Governor    Anderson       62.46       36.34   1.14   

    Rep_Margin  INDEX_NO  
0        34.86        65  
1        31.75        44  
2        51.75        67  
3        46.97        56  
4        64.64         6  
..         ...       ...  
90       41.89        52  
91       53.51        31  
92       39.23        41  
93       46.30        26  
94       26.12        79  

[95 rows x 8 columns]

In [100]:
tn_2016_county_pres_pct.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 95 entries, 0 to 94
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   electdate   95 non-null     datetime64[ns]
 1   county      95 non-null     object        
 2   Republican  95 non-null     float64       
 3   Democratic  95 non-null     float64       
 4   Other       95 non-null     float64       
 5   Rep_Margin  95 non-null     float64       
 6   INDEX_NO    95 non-null     int64         
dtypes: datetime64[ns](1), float64(4), int64(1), object(1)
memory usage: 5.9+ KB


In [102]:
tn_2018_g_county_pct.to_excel(r'../data/2018_election/tn_2018_g_county_pct.xlsx',index=False)